### DISCUSSION TITLE
Universal display: Adding custom ePaper display

### DISCUSSION THREAD

**Discussion by nrother** [arendst/Tasmota#21585] [Support request]
Hi,

I've build a custom PCB hosting a ESP32-C3 and a [Pervasive Display 7.4" eInk Display](https://www.pervasivedisplays.com/product/7-4-e-ink-displays/). I'm trying to use Tasmota to drive the display. As far as I understand the documentation, I should be able to create a custom `display.ini` to control the display.

Unfortunately, [the documentation](https://tasmota.github.io/docs/Universal-Display-Driver/) is quite sketchy... I also looked at [the code](https://github.com/arendst/Tasmota/blob/development/lib/lib_display/UDisplay/uDisplay.cpp), and it seems to contain a lot of special cases for ePaper displays, but I don't really get it.

The screen operates over I2C plus some extra signals. I have some working example code from the vendor that does a full screen update via a (lengthy) sequence of I2C commands. As far I understand the documentation, I need a `:H` section to describe the screen's connections, and then a `:I` section for the initialization? And them a `:f` section for the full screen update? But where do I specify, where the framebuffer content is send?

Do you have any pointers how to proceed? Or would it be the better road to try to incorporate the existing driver into Tasmota?

Side question: Does Tasmota support "red and black"-style ePaper displays?

Thanks for your support!

**Comment by Jason2866** [comment]
The documentation for uDisplay driver is the code ;-)
Only black&white ePaper Displays are supported.

**Comment by s-hadinger** [comment]
I have an M5Stack Paper out of the drawer. I will have a look at epaper displays and possibilities to improve what we currently have
